### PR TITLE
Configurable stage package hash length

### PR DESCRIPTION
**Pull request by johnwparent**
Windows is unable to compile/link paths greater than 260 characters due to limitations in `cl.exe` and `link.exe`, the MSVC compiler and linker respectively. Spack's stage, particularly with a full 32 character hash, runs the risk of pushing packages over that limit (and does for Paraview). This PR adds support for a config option to reduce the size of the stage hash (with a lower limit of 7 characters to maintain uniqueness) and adds a shorter stage root to the default Windows config. 

This PR is designed to be an alternative to #35301

### CODE REVIEW

File: lib/spack/spack/stage.py [hunk 1]
```diff
@@ -150,7 +157,9 @@ def _resolve_paths(candidates):
 
         # Ensure the path is unique per user.
         can_path = sup.canonicalize_path(path)
-        if user not in can_path:
+        # Do not add per user path on Windows - currently only need to support
+        # single user stages
```
Comment (scheibelp) [inline lib/spack/spack/stage.py]: Can you rephrase to provide more context:

> When multiple users share a stage root, we can avoid conflicts between them by adding a per-user subdirectory. Avoid doing this on Windows to keep stage absolute path as short as possible.

File: lib/spack/spack/stage.py [hunk 2]
```diff
@@ -49,6 +50,13 @@
 stage_prefix = "spack-stage-"
 
 
+def compute_stage_name(prefix: str, spec):
```
Comment (scheibelp) [inline lib/spack/spack/stage.py]: The `prefix` parameter isn't used in this method.

I think the parameter should be removed and the `os.path.join` should occur outside of it (which I think already happens in `Stage.__init__`).

File: etc/spack/defaults/windows/config.yaml [hunk 3]
```diff
@@ -3,3 +3,5 @@ config:
   concretizer: clingo
   build_stage::
     - '$spack/.staging'
+    - 'C:/spack'
```
Comment (scheibelp) [inline etc/spack/defaults/windows/config.yaml]: I think this should not be added as a secondary option:

* This list is if earlier entries are inaccessible (i.e. permissions failures)
* I wanted to avoid using this directory without the user's knowledge
* If the prior point is overruled, this should replace the first entry (else the user would still encounter path-length failures).

**Review by scheibelp — CHANGES_REQUESTED** [review]

**Comment by johnwparent** [comment]
@scheibelp rebased again. This PR should effectively be a no-op on non Windows so I'm not sure the failure is originating from this PR.

**Review by scheibelp — APPROVED** [review]
